elytron/src/main/java/org/wildfly/extension/elytron/JdbcRealmDefinition.java

```diff
@@ -165,34 +156,51 @@ public PasswordKeyMapper toPasswordKeyMapper(OperationContext context, ModelNode
                 .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
                 .build();
 
+        static final SimpleAttributeDefinition HASH_ENCODING = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.HASH_ENCODING, ModelType.STRING)
+                .setRequired(false)
+                .setDefaultValue(new ModelNode(BASE64))
+                .setValidator(new StringAllowedValuesValidator(BASE64, HEX))
+                .setAllowExpression(true)
+                .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+                .build();
+
+        static final SimpleAttributeDefinition SALT_ENCODING = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.SALT_ENCODING, ModelType.STRING)
+                .setRequired(false)
+                .setDefaultValue(new ModelNode(BASE64))
+                .setValidator(new StringAllowedValuesValidator(BASE64, HEX))
+                .setAllowExpression(true)
+                .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+                .build();
+
+        @Deprecated
         static final ObjectTypeAttributeDefinition OBJECT_DEFINITION = new ObjectTypeAttributeDefinition.Builder(
                 ElytronDescriptionConstants.BCRYPT_MAPPER, PASSWORD, SALT, ITERATION_COUNT)
                 .setRequired(false)
                 .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
                 .build();
 
-        @Override
-        public ObjectTypeAttributeDefinition getObjectDefinition() {
-            return OBJECT_DEFINITION;
-        }
-
-        @Override
-        public SimpleAttributeDefinition[] getAttributes() {
-            return new SimpleAttributeDefinition[] { PASSWORD, SALT, ITERATION_COUNT};
-        }
+        static final ObjectTypeAttributeDefinition OBJECT_DEFINITION_7_0 = new ObjectTypeAttributeDefinition.Builder(
```

Comment:
Why can't you add the new attributes on `OBJECT_DEFINITION`?

This feels that the "updated" object definition is called OBJECT_DEFINITION_7_0 while the previous one is named `OBJECT_DEFINITION` (and is `@Deprecated`).
Until now we have always had a single object definition in the subsystem and restrict any "versioning" consideration  to transformer and parser code.
I am wondering why did you update the model this way for this feature?

Reply:
In the end I needed to nest this quite deeply due to the hierachy in how the persistent resource definition is pulled together.

I did consider not going for versioned XML parsing but then it would mean configuration valid against one server would be invalid against another even though the declared schema version may be common.

Reply:
@jmesnil do you have any further comment on this one?  

I can swap the names over if that is the main convert but as this PR is now coming up to being four weeks old would be good to get it in.

Reply:
@darranl I asked out of curiosity as I wonder how you will update the code next time you add/modify a parameter. Would you create an `OBJECT_DEFINITION_X`?

My personal preference would be to keep `OBJECT_DEFINITION` for the current definition and renamed the deprecated one to `OBJECT_DEFINITION_6_0`.
However this internal refactoring does not impact the feature from this PR so I'm ok with merging it as it is.